examples/generative/stylegan.py

```diff
@@ -0,0 +1,736 @@
+
+    def gradient_loss(self, grad):
+        loss = tf.square(grad)
+        loss = tf.reduce_sum(loss, axis=np.arange(1, len(tf.shape(loss))))
+        loss = tf.sqrt(loss)
+        loss = tf.reduce_mean(tf.square(loss - 1))
+        return loss
+
+    def train_step(self, real_images):
+
+        self.train_step_counter.assign_add(1)
+
+        if self.phase == Phase.TRANSITION:
+            self.alpha.assign(
+                tf.cast(self.train_step_counter / self.steps_per_epoch, tf.float32)
+            )
+        elif self.phase == Phase.STABLE:
+            self.alpha.assign(1.0)
+        else:
+            raise NotImplementedError
+
+        batch_size = tf.shape(real_images)[0]
+        real_labels = tf.ones(batch_size)
+        fake_labels = -tf.ones(batch_size)
+
+        z = tf.random.normal((batch_size, self.z_dim))
+        const_input = tf.ones(tuple([batch_size] + list(self.g_input_shape)))
+        noise = self.generate_noise(batch_size)
+
+        # generator
+        with tf.GradientTape() as g_tape:
+            w = self.mapping(z)
+            fake_images = self.generator([const_input, w, noise])
+            pred_fake = self.discriminator(fake_images)
+            g_loss = wasserstein_loss(real_labels, pred_fake)
+
+            variables = self.mapping.variables + self.generator.variables
+            gradients = g_tape.gradient(g_loss, variables)
+            self.g_optimizer.apply_gradients(zip(gradients, variables))
+
+        # discriminator
+        with tf.GradientTape() as gradient_tape, tf.GradientTape() as total_tape:
+            # forward pass
+            pred_fake = self.discriminator(fake_images)
+            pred_real = self.discriminator(real_images)
+
+            epsilon = tf.random.uniform((batch_size, 1, 1, 1))
+            interpolates = epsilon * real_images + (1 - epsilon) * fake_images
+            gradient_tape.watch(interpolates)
+            pred_fake_grad = self.discriminator(interpolates)
+
+            # calculate losses
+            loss_fake = wasserstein_loss(fake_labels, pred_fake)
+            loss_real = wasserstein_loss(real_labels, pred_real)
+            loss_fake_grad = wasserstein_loss(fake_labels, pred_fake_grad)
+
+            # gradient penalty
+            gradients_fake = gradient_tape.gradient(loss_fake_grad, [interpolates])
+            gradient_penalty = self.loss_weights[
+                "gradient_penalty"
+            ] * self.gradient_loss(gradients_fake)
+
+            # drift loss
+            all_pred = tf.concat([pred_fake, pred_real], axis=0)
+            drift_loss = self.loss_weights["drift"] * tf.reduce_mean(all_pred ** 2)
+
+            d_loss = loss_fake + loss_real + gradient_penalty + drift_loss
+
+            gradients = total_tape.gradient(d_loss, self.discriminator.variables)
+            self.d_optimizer.apply_gradients(
+                zip(gradients, self.discriminator.variables)
+            )
+
+        return {"d_loss": d_loss, "g_loss": g_loss}
```

Comment:
You should use metric trackers here (in order to display the per-epoch average and not the per-batch current value). See e.g. https://keras.io/examples/generative/dcgan_overriding_train_step/#override-trainstep